Add PubNub BLOCKS management module

parfeon opened the pull request:
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
pubnub_blocks

##### ANSIBLE VERSION
```
ansible 2.2.0.0
  config file = 
  configured module search path = ['module', 'library']
```

##### SUMMARY
This module allow to manage [PubNub BLOCKS](https://www.pubnub.com/products/blocks/) during automated integration process. Module allow to create and / or delete blocks along with their event handlers.  
After block creation and event handlers configuration it is possible to run or stop block. Module usage for different purposes covered in it's header and in this [README](https://github.com/pubnub/pubnub-blocks-ansible/blob/master/README.md#module-usage).
Basic usage to create block and start it:
```
- name: Create block with event handlers
  register: module_cache
  pubnub_blocks:
    email: '{{ email }}'
    pwd: '{{ password }}'
    application: '{{ app_name }}'
    keyset: '{{ keyset_name }}'
    block: 'Test Block 1'
    event_handlers:
      -
        src: 'sources/eh1.js'
        name: 'Event Handler 1'
        channels: 'eh-channel-1'
        event: 'js-before-publish'
      -
        src: 'sources/eh2.js'
        name: 'Event Handler 2'
        channels: 'eh-channel-2'
        event: 'js-before-publish'
- name: Start block
  pubnub_blocks:
    cache: '{{ module_cache | default({}) }}'
    application: '{{ app_name }}'
    keyset: '{{ keyset_name }}'
    block: 'Test Block 1'
    state: start
```
Module doesn't have specific output, because all _changed_ state depends on whether any block property / state has been changed.

parfeon committed:
Added PubNub BLOCKS management module.

parfeon committed:
. increased block start/stop wait to 30 seconds

parfeon committed:
. fixed multi-line documentation

parfeon committed:
. fixed YAML formatting in documentation section

parfeon committed:
. removed multi-line commits
. fixed YAML format in description section

ryansb commented on lib/ansible/modules/cloud/pubnub/pubnub_blocks.py:
```diff
@@ -0,0 +1,2090 @@
+
+        Payload include application-wide information and doesn't depend from particular event handler configuration.
+        :rtype:  dict
+        :return: Initial payload dictionary which can be
```
This behaves exactly as `if not val` would. Can you please remove this function & its use in favor of using `not val` in conditionals?

ryansb commented on lib/ansible/modules/cloud/pubnub/pubnub_blocks.py:
```diff
@@ -0,0 +1,2090 @@
+    required: false
+    default: 'start'
+    choices: ['start', 'stop', 'present', 'absent']
+  block:
```
This should just be `name` since `pubnub_blocks: block=foobar` is a bit redundant.

ryansb commented on lib/ansible/modules/cloud/pubnub/pubnub_blocks.py:
```diff
@@ -0,0 +1,2090 @@
+      - Short block description which will be later visible on admin.pubnub.com. Used only if block doesn't exists and won't change description for existing block.
+    required: false
+    default: 'New block'
+  event_handlers:
```
Can these event handlers be specified in different tasks? For example:

```
- pubnub_block:
    name: foobar
    event_handlers:
      src: filename.js
      channels: namehere
- pubnub_block:
    name: foobar
    event_handlers:
      src: filename2.js
      channels: name2here
```

If that happened, what would the behavior be then? Will both handlers exist and be listening, or will only the second one work?

parfeon replied:
@ryansb to call pubnub block user need to specify authorization credentials or result of previous call of module inside of single play.  
To manage blocks or their event handlers it is required to specify their name, block doesn't rely on order of items in `event_handlers`.  
If user explicitly doesn't request any changes to concrete event handler, nothing will be done. If some event handler for same block in single play will appear twice and both with `state` set to **present**, module will check whether there is difference in how event handler represented in block at this moment and in playbook and send updates if required.

ryansb commented on lib/ansible/modules/cloud/pubnub/pubnub_blocks.py:
```diff
@@ -0,0 +1,2090 @@
+  changes:
+    description:
+      - List of fields which should be changed by block itself (doesn't affect any event handlers).
+      - Possible options for change is C(name) and C(description).
```
Is this key for the purpose of renaming an existing block? Usually how this works is some key is global (maybe name) and then if other parameters are changed (like description) the module will detect that the resource description is out of date and change it, so users don't have to manually tell the module what fields to update.

parfeon replied:
@ryansb I forgot to update this line in docs. Actually it allow to change only name of block (if `changes` dictionary appear in event handlers with name key it will rename event handler as well).  
Because module use `name` of block and it's event handlers to manipulate, it is impossible to specify new name in `name` field of module call. `new_name` - this kind of key doesn't look good as for me, that is why I suggested to add expected `changes` in block separately.

ryansb commented on lib/ansible/modules/cloud/pubnub/pubnub_blocks.py:
```diff
@@ -0,0 +1,2090 @@
+      register: module_cache
+      pubnub_blocks:
+        email: '{{ email }}'
+        pwd: '{{ password }}'
```
Slightly longer key names are ok - in this case, can you use `password` instead since "pwd" can be confused for "print working directory".

ryansb commented on lib/ansible/modules/cloud/pubnub/pubnub_blocks.py:
```diff
@@ -0,0 +1,2090 @@
+    PN_URLENCODE = urllib.parse.urlencode
+except ImportError:
+    import urllib
+    PN_URLENCODE = urllib.urlencode
```
To get urlencode, please use `six.moves` instead.

ryansb replied:
To import urlencode, you can use this line:

```
from ansible.compat.module_utils.six.moves.urllib.parse import urlencode
```
Then for the rest of the file just use `urlencode` as normal.

ryansb commented on lib/ansible/modules/cloud/pubnub/pubnub_blocks.py:
```diff
@@ -0,0 +1,2090 @@
+        """
+        data = self._orig_service_data if not _is_empty(self._orig_service_data) else self._block_data
+
+        return data['name'] if not _is_empty(data) and not _is_empty(data.get('name')) else None
```
Instead, you can use `return data.get('name')` and it will return None if name is undefined.

ryansb reviewed (changes requested):
This feels like it's actually multiple modules, one for the actual block (function) and one for event handlers, so that users can upload code and then have multiple playbooks that hook that code up to an event handler.

ryansb commented:
On the subject of splitting out this module into multiple, there's code sharing allowed in Ansible's `module_utils` directory. You'd just need to make a `module_utils/pubnub/` directory and then you could import those functions/classes across modules. See our [Developing Modules](http://docs.ansible.com/ansible/dev_guide/developing_modules.html) docs for more details.

parfeon commented:
@ryansb I know about `module_utils` but I wasn't able to find any words [Developing Modules](http://docs.ansible.com/ansible/dev_guide/developing_modules.html) guide what devs allowed to place there shared module components.  
There is actually no way to separate module into two: block (to create/delete/change/start/stop block) and event handlers (to create/change/delete handlers) because this two processes can't live without each other. There is no BLOCKS REST API which will allow to upload/register event handlers outside of block and BLOCKS REST API won't allow to start block w/o event handlers in it.  
REST API generate unique integer identifiers of `application`, `keyset`, `block` and `event-handler` so the only way to identify them for module user is to use their name which with API calls will be linked with actual identifiers. `keyset_id` can be fetched only for application with known `app_id`. `block_id` can be received only if we have `keyset_id` and same goes to `handler_id` with `block_id`
REST API which allow to manage event handlers manipulation require: `keyset_id` and `block_id` (in case if this is new event handler) or `handler_id` (if we need to change something about it).  
So it basically impossible to separate them into modules even for attempt to decrease amount of parameters which should be passed to module (for handler we will still need to pass same set of data as for block itself).

parfeon committed:
. updated test script to upload event handler sources if non-local connection used for tests
. changed module configuration parameter from 'block' to 'name' (represent target block name)
. updated module docs regarding 'chanes' configuration parameter. Only block 'name' can be changed using this parameter ('description' can be changed directly in module call configuration)
. changed module configuration parameter from 'pwd' tp 'password'
. migrated 'urlencode' import from try..catch to six module imported from Ansible module utils
. removed custom method which is used to verify whether value is empty

parfeon committed:
^ fixed value for key-path function usage with names of application/keyset/block and handlers.
. reduced unnecessary requests (better pre-procesing of previous results)
. event handler name change won't trigger block restart

mattclay commented:
CI failure due to long lines:

```
2017-02-06 18:51:31 ERROR: PEP 8: lib/ansible/modules/cloud/pubnub/pubnub_blocks.py:52:161: E501 line too long (193 > 160 characters) (current)
2017-02-06 18:51:31 ERROR: PEP 8: lib/ansible/modules/cloud/pubnub/pubnub_blocks.py:75:161: E501 line too long (163 > 160 characters) (current)
2017-02-06 18:51:31 ERROR: PEP 8: lib/ansible/modules/cloud/pubnub/pubnub_blocks.py:81:161: E501 line too long (180 > 160 characters) (current)
2017-02-06 18:51:31 ERROR: PEP 8: lib/ansible/modules/cloud/pubnub/pubnub_blocks.py:85:161: E501 line too long (196 > 160 characters) (current)
2017-02-06 18:51:31 ERROR: PEP 8: lib/ansible/modules/cloud/pubnub/pubnub_blocks.py:86:161: E501 line too long (189 > 160 characters) (current)
2017-02-06 18:51:31 ERROR: PEP 8: lib/ansible/modules/cloud/pubnub/pubnub_blocks.py:98:161: E501 line too long (194 > 160 characters) (current)
2017-02-06 18:51:31 ERROR: PEP 8: lib/ansible/modules/cloud/pubnub/pubnub_blocks.py:193:161: E501 line too long (167 > 160 characters) (current)
```

parfeon committed:
^ fixed too long documentation lines

ryansb commented on lib/ansible/modules/cloud/pubnub/pubnub_blocks.py:
```diff
@@ -0,0 +1,2373 @@
+'''
+
+RETURN = '''
+module_cache:
```
Instead of module_cache, can you name this something like "auth_cache" or "credentials_cache" since it seems to only contain account information?

parfeon replied:
@ryansb it is not only credentials, because there is list of applications under account and each of them has keyset under which blocks is stored. So it is not only authorization cache, but basically account information.

ryansb replied:
Ah, I see. Thanks for clarifying!

parfeon replied:
@ryansb in case if developers will be able to complete all with single call of module - nothing will change, but if multiple calls required (maybe want to launch few blocks or spitted event handlers configuration) this cache will allow to eliminate few requests which at this moment may take more time than we wanted.

ryansb commented on lib/ansible/modules/cloud/pubnub/pubnub_blocks.py:
```diff
@@ -0,0 +1,2373 @@
+import time
+
+
+class PubNubAccount(object):
```
Is there no existing Python library to manage BLOCKS API logins & resources? Ansible is ok with modules that require an external client (such as the `boto3` client for AWS APIs, or `shade` for OpenStack). If there's an equivalent for PubNub it would be great not to duplicate code here.

parfeon replied:
@ryansb we do have _cli_ for blocks, but I don't remember any python modules. From docs I got what it not really acceptable to use dependencies and their use should be justified or module won't be accepted into Ansible modules index.

parfeon replied:
@ryansb we will move client's code into **pip** and leave only code in _main_.

ryansb commented on lib/ansible/modules/cloud/pubnub/pubnub_blocks.py:
```diff
@@ -0,0 +1,2373 @@
+                missed_fields.append('channels') if not channels else None
+                missed_fields.append('code') if not code else None
+                missed_fields.append('event') if not event else None
+                error_message = 'Un
```
Typically the keywords for state we use are `started`, `stopped`, `present`, and `absent` since these are states rather than commands.

parfeon replied:
@ryansb I can't say from github comment to where this comment related. Is it about block 'start'/'stop'? I can rename those if this comment was regarding those _state_ values.

ryansb reviewed (changes requested):
Overall looks good, just some minor wording tweaks & a question about whether there's an API client that can be used instead of embedding one into the module.

parfeon committed:
. moved PubNub BLOCKS client code into separate package under PyPi
. updated module code to utilize new package
. changed block start stop state to: started and started

ryansb commented on lib/ansible/modules/cloud/pubnub/pubnub_blocks.py:
```diff
@@ -0,0 +1,565 @@
+    from pubnub_blocks_client import Account, Application, Keyset, Block
+    from pubnub_blocks_client import EventHandler
+    import pubnub_blocks_client.exceptions as exceptions
+    HAS_PUBNUB_BLOCKS_CLIENT = True
```
Thanks much -- having the SDK as a separate library is great.

parfeon replied:
@ryansb I've done changes in YAML docs. Can you review please? Thanks!

ryansb commented on lib/ansible/modules/cloud/pubnub/pubnub_blocks.py:
```diff
@@ -0,0 +1,565 @@
+      - "Intended block state after event handlers creation / update process
+      will be completed."
+    required: false
+    default: \'started\'
```
Since these docs are YAML, you can remove all the `\'` around the doc values.

ryansb commented on lib/ansible/modules/cloud/pubnub/pubnub_blocks.py:
```diff
@@ -0,0 +1,565 @@
+        keyset: '{{ keyset_name }}'
+        name: '{{ block_name }}'
+        state: present
+    - name: Add '{{ event_handler_1_name }}' handler to '{{ block_name }}'
```
These examples are indented too far, can you move them in 2 spaces?

ryansb reviewed (changes requested):
Code looks good, thanks for using an SDK from PyPi instead. Once these small doc changes are fixed this looks good to go.

parfeon committed:
. small changes in documentation format
